coroutine/exception: make it work with co_await

piodul opened the pull request:
The seastar::coroutine::exception wrapper was introduced in order to
return exceptions from coroutines without throwing. A wrapped exception
is supposed to be co_returned from a coroutine.

Unfortunately, due to limitations of the language, coroutines returning
a void future are not allowed to `co_return anything;`,
only `co_return;`, which prevents usage of the wrapper in coroutines
returning void futures.

This commit introduces the possibility of using the wrapper with
co_await. Co_awaiting a wrapper has the same effect as co_returning it,
but unlike co_return it is also legal to do in coroutines returning a
void future.

A helper function seastar::coroutine::return_exception is introduced
which works similarly to seastar::coroutine::make_exception but has a
name which expresses the intent better when used along with co_await.

avikivity commented on include/seastar/coroutine/exception.hh:
```diff
@@ -23,18 +23,47 @@
+    template<typename U>
+    void await_suspend(SEASTAR_INTERNAL_COROUTINE_NAMESPACE::coroutine_handle<U> hndl) noexcept {
+        hndl.promise().set_exception(std::move(eptr));
+        hndl.destroy();
```
Nice hack. Please adjust the test to confirm that destructors in the coroutine are executed. They absolutely should, but I saw no hint of that in the documentation.

piodul replied:
I modified the tests so that the coroutines allocate an object on the stack which increments a counter in constructor and decrements it on destruction. When I removed `hndl.destroy()`, the assertion `count == 0` failed and ASan reported a memory leak, so it suggests that the destructors are run by `coroutine_handle::destroy`.

I also found _some_ version of the C++20 draft and it mentions that `coroutine_handle::destroy` runs destructors of automatic variables in the coroutine state: [dcl.fct.def.coroutine#11](https://timsong-cpp.github.io/cppwp/n4861/dcl.fct.def.coroutine#11). After `await_ready()` returns false the coroutine is considered suspended, so calling `destroy` here shouldn't be considered as undefined behavior ([expr.await#5.1](https://timsong-cpp.github.io/cppwp/n4861/expr.await#5.1)).

piodul replied:
The idea for the "hack" was heavily inspired by this article: https://ladnir.github.io/blog/2020/03/31/result.html

I'm wondering if it would be possible to extend the idea and devise a wrapper over Seastar futures which allows co_awaiting on futures and propagates exceptions without throwing. Its main drawback would be that it wouldn't honor try..catch blocks, just like coroutine::exception, but it could be useful most of the time. Would something like this accepted?

psarna replied:
Sounds good to me, as a heavy user of `return make_exception_future<T>()`

piodul commented on include/seastar/coroutine/exception.hh:
````diff
@@ -47,11 +76,18 @@ struct exception {
 /// Example usage:
 ///
 /// ```
+/// co_await coroutine::make_exception(std::runtime_error("something failed miserably"));
````
Done, added `seastar::coroutine::return_exception`. I noted in the docstring comment that we suggest using `make_exception` with `co_return` and `return_exception` with `co_await`.

piodul commented on include/seastar/coroutine/exception.hh:
```diff
@@ -47,11 +76,18 @@ struct exception {
\ No newline at end of file
+} // coroutine
+
+static inline auto operator co_await(coroutine::exception ex) noexcept {
```
Done, changed to just inline.

piodul commented:
v2:
- Added `return_exception` helper function,
- Changed `static inline` to just `inline`,
- Adjusted the tests so that they verify that when an exception is propagated using `coroutine::exception`, destructors of the objects stored in the coroutine frame are invoked.

avikivity commented on include/seastar/coroutine/exception.hh:
````diff
@@ -54,4 +84,23 @@ exception make_exception(T&& t) {
+/// co_await coroutine::return_exception(std::runtime_error("something failed miserably"));
+/// ```
+template<typename T>
+exception return_exception(T&& t) {
````
add nodiscard here to prevent `coroutine::return_exception(...);` from succeeding.

The documentation is accurate but not helpful. The goal is to return an exception from a coroutine, the user has no goal to create a coroutine::exception. co_await is required, not recommended. Write the documentation from the user point of view, don't describe the internals.

piodul replied:
Added nodiscard to `return_exception`, improved the docs.

Should I add nodiscard to `make_exception`, too? I was hesitant to do it because technically it would be a backwards-incompatible change, but only for the code which uses `make_exception` incorrectly. Alternatively, I can just mark the `exception` nodiscard, then nodiscard will be unnecessary for `return_exception` and `make_exception`.

avikivity replied:
I think it's (marginally) better to annotate {make,return}_exception. The closer the error location to the caller, the easier it is to understand what's going on.

We don't need to keep compatibility with broken code.

piodul replied:
OK, added `[[nodiscard]]` to `make_exception`, too.

piodul commented:
v3:
- Made the docstrings for `make_exception` and `return_exception` more user-friendly,
- Added a `[[nodiscard]]` attribute to `make_exception` and `return_exception`.
